File: demo/src/main/java/com/example/demo/entity/Category.java
package com.example.demo.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Data;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
public class Category extends Base<String> implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "Category name cannot be null")
    @Size(min = 2, max = 100, message = "Category name should have between 2 and 100 characters")
    @Column(nullable = false)
    private String name;

    private String description;
    private String image;
    private boolean active;
    private boolean deleted;
    private boolean editable;
    private boolean visible;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id")
    private Brand brand;

    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY, cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    @JsonIgnore
    private Set<Product> products = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Category)) return false;
        Category category = (Category) o;
        return id != null && id.equals(category.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }
}

File: camerashop_fe/src/model/Category.ts
class Category {
    id: number;
    name: string; // Category name must be between 2 and 100 characters (validation handled in the UI)
    description?: string; // Optional field
    image?: string; // Optional field
    active: boolean;
    deleted: boolean;
    editable: boolean;
    visible: boolean;
    slug?: string; // Optional field
    status?: number; // Optional field
    products?: number[];

    constructor(id: number, name: string, description: string, image: string, active: boolean, deleted: boolean, editable: boolean, visible: boolean, slug: string, status: number, products: number[]) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.image = image;
        this.active = active;
        this.deleted = deleted;
        this.editable = editable;
        this.visible = visible;
        this.slug = slug;
        this.status = status;
        this.products = products;
    }
}

export default Category;
